File: Android App/app/src/main/java/com/sunilson/pro4/baseClasses/Liveticker.java
package com.sunilson.pro4.baseClasses;

/**
 * @author Linus Weiss
 */

import com.sunilson.pro4.exceptions.LivetickerSetException;

/**
 *
 */
public class Liveticker {

    private String livetickerID;
    private String title;
    private String description;
    private String state;
    private String status;
    private String authorID;
    private String userName;
    private String profilePicture;
    private int commentCount, likeCount, viewerCount;
    private Long stateTimestamp;

    public Liveticker() {
        commentCount = 0;
        likeCount = 0;
        viewerCount = 0;
    }

    public int getViewerCount() {
        return viewerCount;
    }

    public void setViewerCount(int viewerCount) {
        this.viewerCount = viewerCount;
    }

    public String getPrivacy() {
        return privacy;
    }

    public void setPrivacy(String privacy) throws LivetickerSetException {
        if (privacy == null || privacy.isEmpty()) {
            throw new LivetickerSetException("Privacy not valid");
        }
        this.privacy = privacy;
    }

    private String privacy;

    public String getAuthorID() {
        return authorID;
    }

    public void setAuthorID(String authorID) throws LivetickerSetException {
        if (authorID == null || authorID.isEmpty()) {
            throw new LivetickerSetException("Author ID not valid");
        }
        this.authorID = authorID;
    }

    public String getLivetickerID() {
        return this.livetickerID;
    }

    public void setLivetickerID(String livetickerID) throws LivetickerSetException {
        if (livetickerID == null || livetickerID.isEmpty()) {
            throw new LivetickerSetException("Liveticker ID not valid");
        }
        this.livetickerID = livetickerID;
    }

    public String getTitle() {
        return title;
    }

    public void setTitle(String title) throws LivetickerSetException {
        if (title == null || title.isEmpty()) {
            throw new LivetickerSetException("Title is empty!");
        }

        this.title = title;
    }

    public String getDescription() {
        return description;
    }

    public void setDescription(String description) {
        this.description = description;
    }

    public String getState() {
        return state;
    }

    public void setState(String state) throws LivetickerSetException {
        if (state.isEmpty()) {
            throw new LivetickerSetException("Status is not valid");
        }
        this.state = state;
    }

    public int getCommentCount() {
        return commentCount;
    }

    public void setCommentCount(int commentCount) throws LivetickerSetException {
        if (commentCount < 0) {
            throw new LivetickerSetException("Comment Count too low");
        }
        this.commentCount = commentCount;
    }

    public Long getStateTimestamp() {
        return stateTimestamp;
    }

    public void setStateTimestamp(Long stateTimestamp) throws LivetickerSetException {
        if (stateTimestamp < 0) {
            throw new LivetickerSetException("Start Date not valid");
        }
        this.stateTimestamp = stateTimestamp;
    }

    public String getStatus() {
        return status;
    }

    public void setStatus(String status) {
        this.status = status;
    }

    public String getProfilePicture() {
        return profilePicture;
    }

    public void setProfilePicture(String profilePicture) {
        this.profilePicture = profilePicture;
    }

    public String getUserName() {
        return userName;
    }

    public void setUserName(String userName) {
        this.userName = userName;
    }

    public int getLikeCount() {
        return likeCount;
    }

    public void setLikeCount(int likeCount) {
        this.likeCount = likeCount;
    }
}

File: Firebase Cloud Functions/functions/liveticker.js
var utilities = require(__dirname + '/utilities.js');

function Liveticker(data) {
  this.data = data;
  this.data.commentCount = 0;
  this.data.likeCount = 0;
}

Liveticker.prototype.setUsername = function(username) {
  this.data.userName = username;
}

Liveticker.prototype.setProfilePicture = function(url) {
  this.data.profilePicture = url;
}

Liveticker.prototype.setAuthorID = function(id) {
  this.data.authorID = id;
}

Liveticker.prototype.getData = function() {
  return {
    "title": this.data.title,
    "description": this.data.description,
    "authorID": this.data.authorID,
    "privacy": this.data.privacy,
    "state": this.data.state,
    "idState": this.data.idState,
    "status": this.data.status,
    "commentCount": this.data.commentCount,
    "likeCount": this.data.likeCount,
    "stateTimestamp": this.data.stateTimestamp
  }
}

Liveticker.prototype.validate = function() {

  this.data.title = utilities.sanitizeString(this.data.title);
  this.data.description = utilities.sanitizeString(this.data.description);
  this.data.authorID = utilities.sanitizeStringStrict(this.data.authorID);
  this.data.privacy = utilities.sanitizeStringStrict(this.data.privacy);
  this.data.stateTimestamp = utilities.sanitizeNumber(this.data.stateTimestamp);
  this.data.status = utilities.sanitizeString(this.data.status);
  this.data.commentCount = utilities.sanitizeNumber(this.data.commentCount);
  this.data.likeCount = utilities.sanitizeNumber(this.data.likeCount);

  if (this.data.stateTimestamp <= Date.now()) {
    console.log('warn', 'Start Date was too small!');
    this.data.stateTimestamp = Date.now();
    this.data.state = "started";
    this.data.idState = this.data.authorID + "ab";
  } else {
    if (this.data.stateTimestamp > Date.now() + 432000000) {
      console.log('warn', 'Start Date was too high!');
      this.data.stateTimestamp = Date.now() + 432000000;
    }
    this.data.state = "not started";
    this.data.idState = this.data.authorID + "a";
  }

  //Check if required data is correct
  if (!utilities.checkValidString(this.data.title, false, 5, 30)) {
    console.log('error', "Rejected new Liveticker. Reason: Title Invalid");
    return "Title invalid!";
  } else if (!utilities.checkValidString(this.data.description, true, 0, 9999)) {
    console.log('error', "Rejected new Liveticker. Reason: Description Invalid");
    return "Description invalid!";
  } else if (!utilities.checkValidString(this.data.status, true, 0, 140)) {
    console.log('error', "Rejected new Liveticker. Reason: Status Invalid");
    return "Status invalid!";
  } else if (this.data.privacy !== "public" && this.data.privacy !== "private") {
    console.log('error',
      "Rejected new Liveticker. Reason: Privacy not set");
    return "Privacy must be set!";
  } else if (!utilities.checkValidNumber(this.data.stateTimestamp)) {
    console.log('error', "Rejected new Liveticker. Reason: Start Date invalid");
    return "Start Date invalid!";
  } else if (!utilities.checkValidNumber(this.data.commentCount)) {
    console.log('error', "Rejected new Liveticker. Reason: Comment count invalid");
    return "Comment count invalid!";
  } else if (!utilities.checkValidNumber(this.data.likeCount)) {
    console.log('error', "Rejected new Liveticker. Reason: Like count invalid");
    return "Like count invalid!";
  } else {
    return true;
  }
}

module.exports = Liveticker;
